Sources/FoundationEssentials/Calendar/Calendar_Recurrence.swift

```diff
@@ -0,0 +1,835 @@
+        case .hourly:   .hour
+        case .daily:    .day
+        case .weekly:   .weekOfMonth
+        case .monthly:  .month
+        case .yearly:   .year
+        }
+    }
+}
+
+/// The action of a component of the recurrence rule.
+///
+/// Each field in the recurrence rule has a different effect on the output which
+/// is also different for different frequencies. The effect is documented in the
+/// RFC, and copied below for convenience:
+///
+/// | Component\Freq |.minutely|.hourly |.daily  |.weekly |.monthly|.yearly |
+/// |----------------|---------|--------|--------|--------|--------|--------|
+/// |months          |.limit   |.limit  |.limit  |.limit  |.limit  |.expand |
+/// |weeks           |nil      |nil     |nil     |nil     |nil     |.expand |
+/// |daysOfTheYear   |.limit   |.limit  |nil     |nil     |nil     |.expand |
+/// |daysOfTheMonth  |.limit   |.limit  |.limit  |nil     |.expand |.expand |
+/// |weekdays        |.limit   |.limit  |.limit  |.expand |note 1  |note 2  |
+/// |hours           |.limit   |.limit  |.expand |.expand |.expand |.expand |
+/// |minutes         |.limit   |.expand |.expand |.expand |.expand |.expand |
+/// |setPositions    |.limit   |.limit  |.limit  |.limit  |.limit  |.limit  |
+///
+/// - Note 1: Limit if `daysOfTheMonth` is present; otherwise expand
+/// - Note 2: Limit if `daysOfTheYear` or `daysOfTheMonth` is present; otherwise
+///           expand
+enum ComponentAction {
+    /// Replace each date found so far with a list of dates
+    case expand
+    /// Remove some of the dates which have been found so far
+    case limit
+}
+
+extension Calendar {
+    /// A `Sequence` of `Date`s produced by a given recurrence rule
+    @available(FoundationPreview 0.4, *)
+    struct DatesByRecurring : Sendable, Sequence {
+        typealias Element = Date
+        
+        /// The starting point for the recurrence rule
+        let start: Date
+        /// The recurrenece rule
+        let recurrence: RecurrenceRule
+        /// Range in which the search should occur. If `nil`, return all results
+        let range: Range<Date>?
+        
+        init(start: Date, recurrence: RecurrenceRule, range: Range<Date>?) {
+            self.start = start
+            self.recurrence = recurrence
+            self.range = range
+        }
+        
+        struct Iterator: Sendable, IteratorProtocol {
+            /// The starting date for the recurrence
+            let start: Date
+            /// The recurrence rule that should be used for enumeration
+            let recurrence: RecurrenceRule
+            /// The range in which the sequence should produce results
+            let range: Range<Date>?
+            
+            /// Interval containing `range.lowerBound`, that has the same length
+            /// as the frequency interval. This can be used as a lower bound for
+            /// ``nextBaseRecurrenceDate()``
+            let rangeLowerBoundInterval: DateInterval?
```

Comment:
I sort of wish we didn't mix `DateInterval` and `Range<Date>`, but it's not a showstopper.